include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+			result = instr_execute_cos_eval_pdf_BLEND(instr, srcs, params, bsdf_data);
+		} CASE_END
 #endif
 #ifdef OP_BUMPMAP
-	if (op==OP_BUMPMAP) {
-		instr_execute_BUMPMAP(instr, L); // @Crisspl is your L a global or something?
-	} else
+		CASE_BEGIN(op, OP_BUMPMAP) {
+			instr_execute_BUMPMAP(instr, srcs, precomp);
+		} CASE_END
 #endif
 #ifdef OP_SET_GEOM_NORMAL
-	if (op==OP_SET_GEOM_NORMAL) {
-		instr_execute_SET_GEOM_NORMAL(L); // @Crisspl is your L a global or something?
-	} else
+		CASE_BEGIN(op, OP_SET_GEOM_NORMAL) {
+			instr_execute_SET_GEOM_NORMAL(instr, precomp);
+		} CASE_END
 #endif
-	{} //else "empty braces"
+		CASE_OTHERWISE
+		{} //else "empty braces"
+		END_CASES
+	}
 
-	if (op_isBXDForBlend(op)) //OP_BLEND writes pdf as well, but it happens in instr_execute_cos_eval_pdf_BLEND()
-		writeReg(REG_DST(regs)+3u, pdf);
+	if (op_isBXDForCoatOrBlend(op))
+		writeReg(REG_DST(regs), result);
 }
 
-vec3 irr_bsdf_cos_remainder_and_pdf(in instr_stream_t stream, in irr_glsl_BSDFSample s, out float _out_pdf)
+eval_and_pdf_t irr_bsdf_eval_and_pdf(in MC_precomputed_t precomp, in instr_stream_t stream, in uint generator_offset, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache microfacet)
 {
-#ifndef NO_TWOSIDED
-	bool ts = false;
-#endif
+	setCurrInteraction(precomp);
 	for (uint i = 0u; i < stream.count; ++i)
 	{
 		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
-#ifndef NO_TWOSIDED
-		handleTwosided(ts, instr);
+		uint op = instr_getOpcode(instr);
+
+		bool skip = (i == generator_offset);
+		// skip deltas
+#ifdef OP_THINDIELECTRIC
+		skip = skip || (op == OP_THINDIELECTRIC);
+#endif
+#ifdef OP_DELTATRANS
+		skip = skip || (op == OP_DELTATRANS);
+#endif
+
+		if (skip)
+		{
+			//assert(isBXDF(instr));
+			uvec3 regs = instr_decodeRegisters(instr);
+			writeReg(REG_DST(regs), eval_and_pdf_t(0.0));
+		}
+		else
+		{
+			instr_eval_and_pdf_execute(instr, precomp, s, microfacet);
+		}
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			s = irr_glsl_createLightSample(s.L, currInteraction);
+			microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		}
 #endif
-		instr_remainder_and_pdf_execute(instr, s);
 	}
 
 	eval_and_pdf_t eval_and_pdf = readReg4(0u);
-	_out_pdf = eval_and_pdf.w;
-	return eval_and_pdf.rgb/eval_and_pdf.w;
+	return eval_and_pdf;
 }
 
-irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand)
+irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr_stream_t stream, in vec3 rand, out vec3 out_remainder, out float out_pdf, out irr_glsl_AnisotropicMicrofacetCache out_microfacet, out uint out_gen_rnpOffset, out bool out_invalid_microfacet)
 {
 	uint ix = 0u;
 	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
 	uint op = instr_getOpcode(instr);
+	float weight = 1.0;
```

Comment:
instead of weight.. use `out_pdf` right away
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509400689